README
NAME
    Tee - Pure Perl emulation of GNU tee

VERSION
    This documentation refers to version 0.13_51

SYNOPSIS
      # from Perl
      use Tee;
      tee( $command, @files );
     
  # from the command line
      $ cat README.txt | ptee COPY.txt

DESCRIPTION
    The "Tee" distribution provides the ptee program, a pure Perl emulation
    of the standard GNU tool "tee". It is designed to be a
    platform-independent replacement for operating systems without a native
    "tee" program. As with "tee", it passes input received on STDIN through
    to STDOUT while also writing a copy of the input to one or more files.
    By default, files will be overwritten.

    Unlike "tee", "ptee" does not support ignoring interrupts, as signal
    handling is not sufficiently portable.

    The "Tee" module provides a convenience function that may be used in
    place of "system()" to redirect commands through "ptee".

USAGE
  "tee()"
       tee( $command, @filenames );
       tee( $command, \%options, @filenames );

    Executes the given command via "system()", but pipes it through ptee to
    copy output to the list of files. Unlike with "system()", the command
    must be a string as the command shell is used for redirection and
    piping. The return value of "system()" is passed through, but reflects
    the success of the "ptee" command, which isn't very useful.

    The second argument may be a hash-reference of options. Recognized
    options include:

    *   stderr -- redirects STDERR to STDOUT before piping to ptee (default:
        false)

    *   append -- passes the "-a" flag to ptee to append instead of
        overwriting (default: false)

LIMITATIONS
    Because of the way that "Tee" uses pipes, it is limited to capturing a
    single input stream, either STDOUT alone or both STDOUT and STDERR
    combined. A good, portable alternative for capturing these streams from
    a command separately is IPC::Run3, though it does not allow passing it
    through to a terminal at the same time.

SEE ALSO
    *   ptee

    *   IPC::Run3

    *   IO::Tee

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=Tee>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2006-2008 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Perl module Tee

{{$NEXT}}

0.14      2011-01-28 09:40:36 EST5EDT

    [BUG FIXES]

    - fix error writing to undef handle on MSWin32 in some circumstances

    [OTHER]

    - converted distribution to Dist::Zilla

0.13_51 Thu Dec 11 00:48:32 EST 2008

    - moved guts of ptee to Tee::App; removed ptee search logic from Tee.pm

    - bumped some prerequisites; dropped others

    - changed to the Apache License, version 2.0; (it's clearer, relicensable,
      and is explicit about contributions)

0.13 Thu Aug 17 13:09:03 EDT 2006
    - customize Makefile.PL for auto/Tee
    - handle spaces in ptee path in Tee.pm

0.12 Thu Aug 17 07:40:23 EDT 2006
    - store a copy of ptee in auto/Tee for tee()
    - removed unused test files
    - downgraded IPC::Run3 prereq version to 0.033

0.11 Thu Aug  3 14:49:31 EDT 2006
    - added multiple file support to tee() to match documentation
    - Pod addition and cleanups

0.10 Wed Aug  2 22:20:22 EDT 2006
    - initial release



t/Expected.pm
package t::Expected;
@EXPORT = qw( expected );
@ISA = qw( Exporter );
use strict;
use Exporter;

my %expected = (
    "STDOUT" => "# STDOUT: hello world\n",
    "STDERR" => "# STDERR: goodbye, cruel world\n",
);

sub expected {
    return $expected{+shift};
}

1;


t/fatality.pl
select(STDERR);
$|++;
select(STDOUT);
$|++;
print STDOUT "# STDOUT: hello world\n";
print STDERR "# STDERR: goodbye, cruel world\n";
exit 1;


t/helloworld.pl
select(STDERR);
$|++;
select(STDOUT);
$|++;
print STDOUT "# STDOUT: hello world\n";
print STDERR "# STDERR: goodbye, cruel world\n";


lib/Tee.pm
use strict;
use warnings;
package Tee;
# ABSTRACT: Pure Perl emulation of GNU tee

use Exporter ();
use Probe::Perl;

our @ISA         = qw (Exporter);
our @EXPORT      = qw (tee);

#--------------------------------------------------------------------------#
# Platform independent ptee invocation
#--------------------------------------------------------------------------#

my $perl = Probe::Perl->find_perl_interpreter;
my $ptee_cmd = "$perl -MTee::App -e run --";

#--------------------------------------------------------------------------#
# Functions
#--------------------------------------------------------------------------#

sub tee {
    my $command = shift;
    my $options;
    $options = shift if (ref $_[0] eq 'HASH');
    my $files = join(" ", @_);
    my $redirect = $options->{stderr} ? " 2>&1 " : q{};
    my $append = $options->{append} ? " -a " : q{};
    system( "$command $redirect | $ptee_cmd $append $files" );
}

1; # modules must be true

__END__
#--------------------------------------------------------------------------#
# main pod documentation 
#--------------------------------------------------------------------------#

=begin wikidoc

= SYNOPSIS

 # from Perl
 use Tee;
 tee( $command, @files );
 
 # from the command line
 $ cat README.txt | ptee COPY.txt

= DESCRIPTION

The {Tee} distribution provides the [ptee] program, a pure Perl emulation of
the standard GNU tool {tee}.  It is designed to be a platform-independent
replacement for operating systems without a native {tee} program.  As with
{tee}, it passes input received on STDIN through to STDOUT while also writing a
copy of the input to one or more files.  By default, files will be overwritten.

Unlike {tee}, {ptee} does not support ignoring interrupts, as signal handling
is not sufficiently portable.

The {Tee} module provides a convenience function that may be used in place of
{system()} to redirect commands through {ptee}. 

= USAGE

== {tee()}

  tee( $command, @filenames );
  tee( $command, \%options, @filenames );

Executes the given command via {system()}, but pipes it through [ptee] to copy
output to the list of files.  Unlike with {system()}, the command must be a
string as the command shell is used for redirection and piping.  The return
value of {system()} is passed through, but reflects the success of 
the {ptee} command, which isn't very useful.

The second argument may be a hash-reference of options.  Recognized options
include:

* stderr -- redirects STDERR to STDOUT before piping to [ptee] (default: false)
* append -- passes the {-a} flag to [ptee] to append instead of overwriting
(default: false)

= LIMITATIONS

Because of the way that {Tee} uses pipes, it is limited to capturing a single
input stream, either STDOUT alone or both STDOUT and STDERR combined.  A good,
portable alternative for capturing these streams from a command separately is
[IPC::Run3], though it does not allow passing it through to a terminal at the
same time.

= SEE ALSO

* [ptee]
* IPC::Run3
* IO::Tee

= BUGS

Please report any bugs or feature using the CPAN Request Tracker.  
Bugs can be submitted through the web interface at 
[http://rt.cpan.org/Dist/Display.html?Queue=Tee]

When submitting a bug or request, please include a test-file or a patch to an
existing test-file that illustrates the bug or desired feature.

=end wikidoc

=cut


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.30
use ExtUtils::MakeMaker;
WriteMakefile
(
          'PL_FILES' => {},
          'INSTALLDIRS' => 'site',
          'NAME' => 'Tee',
          'EXE_FILES' => [
                           'bin/ptee'
                         ],
          'VERSION_FROM' => 'lib/Tee.pm',
          'PREREQ_PM' => {
                           'Getopt::Long' => '2.2',
                           'IO::File' => 0,
                           'IO::CaptureOutput' => '1.09',
                           'File::Spec' => '3.27',
                           'Exporter' => 0,
                           'Test::More' => '0.47',
                           'Probe::Perl' => '0.01',
                           'File::Temp' => '0.2',
                           'File::Basename' => 0
                         }
        )
;


lib/Tee/App.pm
use strict;
use warnings;
package Tee::App;
# ABSTRACT: Implementation of ptee

use Exporter ();
use File::Basename qw/basename/;
use Getopt::Long;
use IO::File;
our @ISA = 'Exporter';
our @EXPORT = qw/run/;

#--------------------------------------------------------------------------#
# define help text
#--------------------------------------------------------------------------#

my $help_text = <<'END_HELP';

 ptee [OPTIONS]... [FILENAMES]...

 OPTIONS:
 
    --append or -a
        append to file(s) rather than overwrite

    --help or -h
        give usage information

    --version or -V
        print the version number of this program

END_HELP

$help_text =~ s/\A.+?( ptee.*)/$1/ms;

sub run {

  #--------------------------------------------------------------------------#
  # process command line
  #--------------------------------------------------------------------------#

  my %opts;
  GetOptions( \%opts,
      'version|V',
      'help|h|?',
      'append|a',
  );

  #--------------------------------------------------------------------------#
  # options
  #--------------------------------------------------------------------------#

  if ($opts{version}) {
      print basename($0), " $main::VERSION\n";
      exit 0;
  }

  if ($opts{help}) {
      print "Usage:\n$help_text";
      exit 1;
  }

  my $mode = $opts{append} ? ">>" : ">";

  #--------------------------------------------------------------------------#
  # Setup list of filehandles
  #--------------------------------------------------------------------------#

  my @files;

  for my $file ( @ARGV ) {
      my $f = IO::File->new("$mode $file") 
          or die "Could't open '$file' for writing: $!'";
      push @files, $f;
  }

  #--------------------------------------------------------------------------#
  # Tee input to the filehandle list
  #--------------------------------------------------------------------------#

  my $buffer_size = 1024;
  my $buffer;

  while ( sysread( STDIN, $buffer, $buffer_size ) > 0 ) {
      syswrite STDOUT, $buffer;
      for my $fh ( @files ) {
          syswrite $fh, $buffer;
      }
  }
  return;
}

1;

__END__

=for Pod::Coverage run

=begin wikidoc

= DESCRIPTION

Guts of the {ptee} command.

= SEE ALSO

* [ptee]

= BUGS

Please report any bugs or feature using the CPAN Request Tracker.  
Bugs can be submitted through the web interface at 
[http://rt.cpan.org/Dist/Display.html?Queue=Tee]

When submitting a bug or request, please include a test-file or a patch to an
existing test-file that illustrates the bug or desired feature.

=end wikidoc

=cut


Build.PL
# Copyright (c) 2008 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use 5.006;
use strict;
use warnings;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'Tee',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
      'Exporter' => 0,
      'File::Basename' => 0,
      'Getopt::Long' => 2.2,  
      'IO::File' => 0,
      'Probe::Perl' => 0.01,
    },
    build_requires => {
      'File::Spec' => 3.27, 
      'File::Temp' => 0.20, 
      'IO::CaptureOutput' => 1.09,
      'Test::More' => 0.47, # 0.47 in perl 5.006002
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t /],
        } 
    },
);
$build->create_build_script;
